src/staking/UniswapV3Staking.sol

```diff
@@ -0,0 +1,130 @@
+    event RewardClaimed(address indexed user, uint256 reward);
+
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor() { _disableInitializers(); }
+
+    function initialize(
+        address _owner, 
+        IERC20 _rewardsToken,
+        INonfungiblePositionManager _positionManager,
+        IDyadXP _dyadXP,
+        DNft _dnft, 
+        uint256 _rewardsRate,
+        address _rewardsTokenHolder
+    ) public initializer {
+        __UUPSUpgradeable_init();
+        __Ownable_init(msg.sender);
+
+        rewardsToken = _rewardsToken;
+        positionManager = _positionManager;
+        dyadXP = _dyadXP; 
+        dnft = _dnft;
+        rewardsRate = _rewardsRate;
+        rewardsTokenHolder = _rewardsTokenHolder;
+    }
+
+    function stake(uint256 noteId, uint256 tokenId) external {
+        require(dnft.ownerOf(noteId) == msg.sender, "You are not the Note owner");
+
+        StakeInfo storage stakeInfo = stakes[noteId];
+        require(!stakeInfo.isStaked, "Note already used for staking"); 
+
+        (,,,,,,, uint128 liquidity,,,,) = positionManager.positions(tokenId);
+        require(liquidity > 0, "No liquidity");
+
+        positionManager.safeTransferFrom(msg.sender, address(this), tokenId);
+
+        stakes[noteId] = StakeInfo({
+          liquidity: liquidity,
+          lastRewardTime: block.timestamp,
+          tokenId: tokenId,
+          isStaked: true
+        });
+
+        emit Staked(msg.sender, noteId, tokenId, liquidity);
+    }
+
+    function unstake(uint256 noteId, address recipient) external {
+        require(dnft.ownerOf(noteId) == msg.sender, "You are not the Note owner");
+        StakeInfo storage stakeInfo = stakes[noteId];
+
+        _claimRewards(noteId, stakeInfo, recipient);
+
+        positionManager.safeTransferFrom(address(this), msg.sender, stakeInfo.tokenId);
+
+        delete stakes[noteId];
+
+        emit Unstaked(msg.sender, noteId, stakeInfo.tokenId);
+    }
+
+    function claimRewards(uint256 noteId, address recipient) external {
+        StakeInfo storage stakeInfo = stakes[noteId];
+        require(dnft.ownerOf(noteId) == msg.sender, "You are not the Note owner");
+
+        _claimRewards(noteId, stakeInfo, recipient);
+    }
+
+    function _claimRewards(uint256 noteId, StakeInfo storage stakeInfo, address recipient) internal {
+        uint256 rewards = _calculateRewards(noteId, stakeInfo);
+        stakeInfo.lastRewardTime = block.timestamp;
+
+        if (rewards > 0) {
+            rewardsToken.transferFrom(rewardsTokenHolder, recipient, rewards);
+            emit RewardClaimed(recipient, rewards);
+        }
+    }
+
+    function _calculateRewards(uint256 noteId, StakeInfo storage stakeInfo) internal view returns (uint256) {
+        uint256 timeDiff = block.timestamp - stakeInfo.lastRewardTime;
+
+        uint256 xp = dyadXP.balanceOfNote(noteId); 
+
+        return timeDiff * rewardsRate * stakeInfo.liquidity * xp;
```

Comment:
do we need a scaling factor here? Should this be based on liquidity in comparison to total liquidity? should xp scaling be based on total XP?

Reply:
Also, it doesn't seem like this is accounting for out of range liquidity at all? Just total liquidity?

Reply:
we can't use totalXP because we would run out of gas. which scaling factor would you use?

Good point, yeah it only considers total liquidity.